### DISCUSSION TITLE
$id URI which are not URLs

### DISCUSSION THREAD

**Discussion by danielmarschall** [json-schema-org/community#205] [Q&A] [answered]
Hello,

I have read at https://json-schema.org/understanding-json-schema/structuring.html#id , that `$id` needs to be an URI.

However, I believe that the author's intention was that it is an absolute URL  (i.e. enforcing `http://www.` instead of using `www.`) . This becomes clear to me because the author also wrote "_this is analogous to the `<base>` HTML tag_".

A URI can also be a URN, for example `urn:oid:2.999` or `urn:ietf:rfc:1234`.

My question is: Is this OK to use a URN , or will this cause problems with JSON validators? ([possible example](https://github.com/jupyter/telemetry/issues/56))

**The background of my question:**
I am currently developing an [Internet Draft](https://datatracker.ietf.org/doc/draft-viathinksoft-oidip/), and I want to make everything as perfect as possible (after all, if it becomes an RFC, it can never be changed again).
Since I have no permanent place where the file is hosted (and an RFC should not rely on a file hosted somewhere), my idea was to use a URN instead of a URL.

In my example it would be:
`"$id": "urn:ietf:id:draft-viathinksoft-oidip-04"`

This way, the JSON schema can be included in the Internet Draft without the requirement that it is hosted somewhere.

If this is _not_ what the author has intended, what is your idea to solve the problem? Also, in this case it would be good to update the specification to clarify that the URI must not be a URN.

Thank you for your help and ideas.

Kind regards,
Daniel Marschall

**Comment by gregsdennis** [comment]
> This way, the JSON schema can be included in the Internet Draft without the requirement that it is hosted somewhere.

It definitely doesn't need to be hosted somewhere.

"Note that this URI is an identifier and not necessarily a network locator. In the case of a network-addressable URL, a schema need not be downloadable from its canonical URI." - Core spec, section [8.2.1](https://json-schema.org/draft/2020-12/json-schema-core.html#name-the-id-keyword)

I believe that URNs are fine, but I've requested a second opinion from the team.

Interestingly, we don't have any tests around URNs in `$id` in the test suite, so implementation support may be spotty.

**Comment by handrews** [comment]
[Appendex A](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-schema-identification-examp) includes an example with `"$id": "urn:uuid:ee564b8a-7a87-4125-8c96-e9f123d6766f"`, so yes, URNs are expected to work.

I have also used [tag: URIs](https://datatracker.ietf.org/doc/html/rfc4151) successfully with more than one validator.

**Comment by jdesrosiers** [comment, marked as answer]
I suggest reading the section that comes before the one you linked called [Schema identification](https://json-schema.org/understanding-json-schema/structuring.html#schema-identification). Common use of URI terminology doesn't always match up with the way the terms are used in [RFC-3986](https://datatracker.ietf.org/doc/html/rfc3986/), which is why I define how I use those terms before I used them. The wording used to describe `$id` is correct and I think once you're aligned with how those terms are being used, it will make more sense.

There's also a box in that section that let's you know that URIs in JSON Schema are only identifiers. They are not intended to be fetched over HTTP. So, feel free to use an http scheme URI in your I-D if you wish, but URNs are fine too if you prefer.

**Comment by danielmarschall** [comment]
Hello,

thank you very much for your detailed and helpful answers!

Also, I'm happy to see that the JSON schema is also available as Internet Draft, and I am looking forward to this upcoming RFC :-)

Kind regards,
Daniel Marschall
